controllers/dbuser_controller.go

```diff
@@ -0,0 +1,344 @@
+		return r.manageError(ctx, dbucr, err, false)
+	}
+
+	// Check if DbUser is marked to be deleted
+	ownership := []metav1.OwnerReference{}
+	ownership = append(ownership, metav1.OwnerReference{
+		APIVersion: dbucr.APIVersion,
+		Kind:       dbucr.Kind,
+		Name:       dbucr.Name,
+		UID:        dbucr.GetUID(),
+	},
+	)
+
+	engine, err := dbcr.GetEngineType()
+	if err != nil {
+		return r.manageError(ctx, dbucr, err, false)
+	}
+
+	userSecret, err := r.getDbUserSecret(ctx, dbucr)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			dbName := fmt.Sprintf("%s-%s", dbucr.Namespace, dbucr.Spec.DatabaseRef)
+			secretData, err := generateDatabaseSecretData(dbucr.ObjectMeta, engine, dbName)
+			if err != nil {
+				logrus.Errorf("can not generate credentials for database - %s", err)
+				return r.manageError(ctx, dbucr, err, false)
+			}
+			newDbUserSecret := kci.SecretBuilder(dbucr.Spec.SecretName, dbucr.Namespace, secretData, ownership)
+			err = r.Create(ctx, newDbUserSecret)
+			if err != nil {
+				// failed to create secret
+				return r.manageError(ctx, dbucr, err, false)
+			}
+			userSecret = newDbUserSecret
+		} else {
+			logrus.Errorf("could not get database secret - %s", err)
+			return r.manageError(ctx, dbucr, err, true)
+		}
+	}
+
+	creds, err := parseDbUserSecretData(dbcr.Status.InstanceRef.Spec.Engine, userSecret.Data)
+	if err != nil {
+		return r.manageError(ctx, dbucr, err, false)
+	}
+
+	if isDbUserChanged(dbucr, userSecret) {
+		dbucr.Status.Status = false
+	}
+
+	db, dbuser, err := determinDatabaseType(dbcr, creds)
+	if err != nil {
+		// failed to determine database type
+		return r.manageError(ctx, dbucr, err, false)
+	}
+	adminSecretResource, err := r.getAdminSecret(ctx, dbcr)
+	if err != nil {
+		// failed to get admin secret
+		return r.manageError(ctx, dbucr, err, false)
+	}
+
+	adminCred, err := db.ParseAdminCredentials(adminSecretResource.Data)
+	if err != nil {
+		// failed to parse database admin secret
+		return r.manageError(ctx, dbucr, err, false)
+	}
+
+	dbuser.AccessType = dbucr.Spec.AccessType
+	dbuser.Password = creds.Password
+	dbuser.Username = fmt.Sprintf("%s-%s", dbucr.GetObjectMeta().GetNamespace(), dbucr.GetObjectMeta().GetName())
+
+	if dbucr.GetDeletionTimestamp() != nil {
+		dbucr.Status.Phase = DBUCR_PHASE_DELETING
+		if containsString(dbucr.ObjectMeta.Finalizers, "dbuser."+dbucr.Name) {
+			if err := database.DeleteUser(db, dbuser, adminCred); err != nil {
+				logrus.Errorf("DBUser: namespace=%s, name=%s failed deleting a user - %s", dbucr.Namespace, dbucr.Name, err)
+				return r.manageError(ctx, dbucr, err, false)
+			}
+			kci.RemoveFinalizer(&dbucr.ObjectMeta, "dbuser."+dbucr.Name)
+			err = r.Update(ctx, dbucr)
+			if err != nil {
+				logrus.Errorf("error resource updating - %s", err)
+				return r.manageError(ctx, dbucr, err, false)
+			}
+			kci.RemoveFinalizer(&dbcr.ObjectMeta, "dbuser."+dbucr.Name)
```

Comment:
duplicated

Reply:
It would work for ``ObjectMeta`,  but I'm not sure how to make it nice on the `Update` object function

Reply:
Think could be done with golang generics. but think for 2 cases this is currently fine.